File: src/main/java/com/rafaelfiume/tictactoe/Game.java
package com.rafaelfiume.tictactoe;


public interface Game {

    ////////////////////////
    //        Turn        //
    ////////////////////////

    Player currentPlayer();

    ////////////////////////
    //        Game        //
    ////////////////////////

    Game snapshot();

    boolean gameIsOn();

    boolean gameHasNotStarted();

    boolean hasWinner();

    boolean hasDraw();

    Player winner();

    ////////////////////////
    //        Board       //
    ////////////////////////

    void playerChooses(BoardPosition boardPosition);

    char topLeft();
    char topCenter();
    char topRight();
    char midLeft();
    char center();
    char midRight();
    char bottomLeft();
    char bottomCenter();
    char bottomRight();

}

File: src/main/java/com/rafaelfiume/tictactoe/Game.kt
package com.rafaelfiume.tictactoe


interface Game {

    fun playerChooses(boardPosition: BoardPosition)

    fun gameIsOn(): Boolean

    fun gameHasNotStarted(): Boolean

    fun winner(): Player

    fun hasWinner(): Boolean

    fun hasDraw(): Boolean

    fun snapshot(): Game

    // Board
    fun board(): Board

    // Turn
    fun currentPlayer(): Player

}
